staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```diff
@@ -247,16 +253,46 @@ func (s *EtcdOptions) ApplyWithStorageFactoryTo(factory serverstorage.StorageFac
 		return fmt.Errorf("EtcdOptions.Apply called without completion")
 	}
 
-	if !s.SkipHealthEndpoints {
+	if s.dynamicTransformers != nil {
+		factory = &transformerStorageFactory{
+			delegate:            factory,
+			dynamicTransformers: s.dynamicTransformers,
+		}
+	}
+
+	if !s.AvoidRewire {
 		if err := s.addEtcdHealthEndpoint(c); err != nil {
 			return err
 		}
-	}
 
-	if len(s.transformerOverrides) > 0 {
-		factory = &transformerStorageFactory{
-			delegate:             factory,
-			transformerOverrides: s.transformerOverrides,
+		// register kms encryption config hot reload controller if dynamicTransformers are not nil
+		if s.dynamicTransformers != nil {
+			err := c.AddPostStartHook(
+				"start-kms-encryption-config-hot-reloader",
+				func(hookContext server.PostStartHookContext) error {
+					// get context
+					// we ignore the cancel func because this context should only be canceled when stopCh is closed
+					ctx, _ := wait.ContextForChannel(hookContext.StopCh)
+
+					kmsConfigController, err := kmsconfigcontroller.NewDynamicKMSEncryptionConfiguration(
```

Comment:
Hmm lets move this `New` call into `.Complete`, and just have this hook call `.Run`.  We also need to come up with a way to ensure that `.Run` only happens once because `dynamicTransformers` relies on that since it is not safe to have multiple callers to `.Set`

Reply:
Maybe we would call `.Run` in `.Complete` - that seems wrong because that is what hooks are supposed to be for?  Or set up the hook in there by passing `c.AddPostStartHook` as a param?